sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -310,13 +310,16 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
 
   protected def ansiEnabled: Boolean
 
+  protected def withDataType(dataType: DataType): CastBase
+
```

Comment:
Looks like the implementation is the same between all subclasses, why can't we implement it here?

Reply:
The `copy` methods are not generated for abstract classes.